enterprise/frontend/src/metabase-enterprise/data-studio/data-model/components/TablePicker/components/TablePickerTreeTable.tsx

```diff
@@ -0,0 +1,553 @@
+  useRef,
+} from "react";
+import { t } from "ttag";
+
+import { useListUsersQuery } from "metabase/api";
+import { useNumberFormatter } from "metabase/common/hooks/use-number-formatter";
+import type { SelectionState, TreeTableColumnDef } from "metabase/ui";
+import {
+  Box,
+  EntityNameCell,
+  Icon,
+  TreeTable,
+  useTreeTableInstance,
+} from "metabase/ui";
+import type { TableId, UserId } from "metabase-types/api";
+
+import { useSelection } from "../../../pages/DataModel/contexts/SelectionContext";
+import {
+  type NodeSelection,
+  getSchemaId,
+  isItemSelected,
+  toggleDatabaseSelection,
+  toggleSchemaSelection,
+} from "../bulk-selection.utils";
+import { TYPE_ICONS } from "../constants";
+import type {
+  ChangeOptions,
+  DatabaseNode,
+  RootNode,
+  SchemaNode,
+  TableNode,
+  TablePickerTreeNode,
+  TreePath,
+} from "../types";
+import { nodeToTreePath, transformToTreeTableFormat } from "../utils";
+
+type OriginalNode = DatabaseNode | SchemaNode | TableNode;
+
+interface Props {
+  tree: RootNode;
+  path: TreePath;
+  isExpanded: (key: string) => boolean;
+  onToggle: (key: string, value?: boolean) => void;
+  onChange?: (path: TreePath, options?: ChangeOptions) => void;
+  reload?: (path: TreePath) => void;
+}
+
+export function TablePickerTreeTable({
+  tree,
+  path,
+  isExpanded,
+  onToggle,
+  onChange,
+  reload,
+}: Props) {
+  const {
+    selectedTables,
+    setSelectedTables,
+    selectedSchemas,
+    setSelectedSchemas,
+    selectedDatabases,
+    setSelectedDatabases,
+    selectedItemsCount,
+  } = useSelection();
+
+  const numberFormatOptions = useMemo(() => ({ maximumFractionDigits: 0 }), []);
+  const formatNumber = useNumberFormatter(numberFormatOptions);
+  const lastSelectedTableIndex = useRef<number | null>(null);
+  const instanceRef = useRef<ReturnType<
+    typeof useTreeTableInstance<TablePickerTreeNode>
+  > | null>(null);
+
+  const { data: usersData } = useListUsersQuery();
+  const ownerNameById = useMemo(() => {
+    const users = usersData?.data ?? [];
+    return new Map<UserId, string>(
+      users.map((user) => [user.id, user.common_name]),
+    );
+  }, [usersData]);
+
+  const treeData = useMemo(() => transformToTreeTableFormat(tree), [tree]);
+
+  const nodeKeyToId = useMemo(() => {
+    const map = new Map<string, string>();
+    const traverse = (nodes: TablePickerTreeNode[]) => {
+      for (const node of nodes) {
+        map.set(node.nodeKey, node.id);
+        if (node.children) {
+          traverse(node.children);
+        }
+      }
+    };
+    traverse(treeData);
+    return map;
+  }, [treeData]);
+
+  const nodeKeyToOriginal = useMemo(() => {
+    const map = new Map<string, OriginalNode>();
+    for (const db of tree.children) {
+      map.set(db.key, db);
+      for (const schema of db.children) {
+        map.set(schema.key, schema);
+        for (const table of schema.children) {
+          map.set(table.key, table);
+        }
+      }
+    }
+    return map;
```

Comment:
nested loops like these give me shivers :fearful: 
Can we use something a little more friendly like:
```suggestion
    const dbs = tree.children;
    const schemas = dbs.flatMap((db) => db.children);
    const tables = schemas.flatMap((schema) => schema.children);

    return new Map([...dbs, ...schemas, ...tables].map((n) => [n.key, n]));
```